File: server/src/main/java/de/ingrid/igeserver/services/KeycloakMockService.java
package de.ingrid.igeserver.services;

import de.ingrid.igeserver.model.User;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.keycloak.adapters.springsecurity.token.KeycloakAuthenticationToken;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.context.annotation.Profile;
import org.springframework.stereotype.Service;

import java.security.Principal;
import java.util.*;

@Service
@Profile("dev")
public class KeycloakMockService implements UserManagementService {

    private Logger log = LogManager.getLogger(KeycloakMockService.class);

    @Value("${dev.user.roles:}")
    String[] mockedUserRoles;

    @Value("${dev.user.login:}")
    String mockedLogin;

    @Value("${dev.user.firstName:}")
    String mockedFirstName;

    @Value("${dev.user.lastName:}")
    String mockedLastName;


    public List<User> getUsers(Principal principal) {

        List<User> mockUsers = new ArrayList<>();
        User user = new User();
        user.setLogin(mockedLogin);
        user.setFirstName(mockedFirstName);
        user.setLastName(mockedLastName);
        mockUsers.add(user);
        return mockUsers;
    }

    @Override
    public Set<String> getRoles(KeycloakAuthenticationToken principal) {
        return new HashSet<>(Arrays.asList(mockedUserRoles));
    }

    @Override
    public String getName(KeycloakAuthenticationToken principal) {
        return "Michael Mustermann";
    }


    @Override
    public User getUser(Principal principal, String login) {
        User user = new User();
        user.setLogin("admin");
        user.setFirstName("Mocked");
        user.setLastName("User");
        return user;
    }

}

File: server/src/main/java/de/ingrid/igeserver/services/KeycloakMockService.kt
package de.ingrid.igeserver.services

import de.ingrid.igeserver.model.User
import org.apache.logging.log4j.LogManager
import org.keycloak.adapters.springsecurity.token.KeycloakAuthenticationToken
import org.springframework.beans.factory.annotation.Value
import org.springframework.context.annotation.Profile
import org.springframework.stereotype.Service
import java.io.Closeable
import java.security.Principal
import java.util.*

class DummyClient : Closeable {
    override fun close() {
        // do nothing
    }
}

@Service
@Profile("dev")
class KeycloakMockService : UserManagementService {
    private val log = LogManager.getLogger(KeycloakMockService::class.java)

    @Value("\${dev.user.roles:}")
    lateinit var mockedUserRoles: Array<String>

    @Value("\${dev.user.login:}")
    lateinit var mockedLogin: String

    @Value("\${dev.user.firstName:}")
    lateinit var mockedFirstName: String

    @Value("\${dev.user.lastName:}")
    lateinit var mockedLastName: String

    override fun getUsersWithIgeRoles(principal: Principal?): Set<User> {
        val mockUsers: MutableList<User> = ArrayList()
        val user =
            User(mockedLogin, mockedFirstName, mockedLastName, "", "", "", "", emptyList(), Date(0), Date(0), Date(0))
        mockUsers.add(user)
        return mockUsers.toSet()
    }

    override fun getUsers(principal: Principal?): Set<User> {
        TODO("Not yet implemented")
    }

    override fun getLatestLoginDate(principal: Closeable, login: String): Date? {
        return Date()
    }

    override fun getRoles(principal: KeycloakAuthenticationToken?): Set<String> {
        return HashSet(Arrays.asList(*mockedUserRoles))
    }

    override fun getName(principal: KeycloakAuthenticationToken?): String {
        return "$mockedFirstName $mockedLastName"
    }

    override fun getClient(principal: Principal?): Closeable {
        return DummyClient()
    }

    override fun getUser(principal: Closeable, login: String): User {
        return User(
            mockedLogin,
            mockedFirstName,
            mockedLastName,
            "",
            "",
            "",
            "",
            emptyList(),
            Date(0),
            Date(0),
            Date(0)
        )
    }

    override fun getCurrentPrincipal(): Principal? {
        return Principal { mockedLogin }
    }

    override fun userExists(principal: Principal, userId: String): Boolean {
        TODO("Not yet implemented")
    }

    override fun createUser(principal: Principal, user: User) {
        TODO("Not yet implemented")
    }

    override fun requestPasswordChange(principal: Principal?, id: String) {
        TODO("Not yet implemented")
    }

    override fun removeRoles(principal: Principal?, userId: String, roles: List<String>) {
        TODO("Not yet implemented")
    }

    override fun addRoles(principal: Principal?, userLogin: String, roles: List<String>) {
        TODO("Not yet implemented")
    }

    override fun updateUser(principal: Principal?, user: User) {
        // do nothing
        log.info("Mocked update user function")
    }
}
